# Functions

Function dalam Python adalah blok kode yang dibuat untuk melakukan tugas tertentu. Mereka memungkinkan Anda untuk mengorganisir kode Anda ke dalam bagian-bagian kecil yang dapat digunakan kembali, sehingga memudahkan dalam mengelola dan memperbaiki program Anda.

In [19]:
# Define the function
def add_three(input_var):
    output_var = input_var + 3
    return output_var


Code di atas adalah sebuah fungsi Python yang disebut add_three. Fungsi ini mengambil satu argumen yang disebut input_var, kemudian menambahkan nilai 3 ke input_var tersebut. Hasil penjumlahan disimpan dalam variabel output_var dan kemudian dikembalikan sebagai hasil dari fungsi menggunakan pernyataan return

## How to run (or "call") a function

In [20]:
# Run the function with 10 as input
new_number = add_three(10)

# Check that the value is 13, as expected
print(new_number)

13


Code di atas memanggil fungsi add_three dengan argumen 10 dan menyimpan hasilnya dalam variabel new_number. Kemudian, nilai dari new_number dicetak menggunakan pernyataan print. Karena add_three(10) akan mengembalikan nilai 10 + 3 = 13, hasil pencetakan akan menjadi 13.

## Examples of more complex functions

In [21]:
def get_pay(num_hours):
    # Pre-tax pay, based on receiving $15/hour
    pay_pretax = num_hours * 15
    # After-tax pay, based on being in 12% tax bracket
    pay_aftertax = pay_pretax * (1 - .12)
    return pay_aftertax

Fungsi di atas adalah `get_pay`, yang mengambil satu argumen yang disebut `num_hours`. Fungsi ini menghitung pembayaran sebelum pajak dan setelah pajak berdasarkan jumlah jam kerja yang diberikan.

1. `pay_pretax` dihitung dengan mengalikan jumlah jam (`num_hours`) dengan tarif upah per jam ($15).
2. `pay_aftertax` dihitung dengan mengalikan `pay_pretax` dengan nilai (1 - 0.12), yang mewakili 1 minus tarif pajak (dalam bentuk desimal).

Setelah itu, fungsi mengembalikan nilai `pay_aftertax`. Jadi, jika Anda memberikan jumlah jam kerja, fungsi akan mengembalikan jumlah pembayaran setelah pajak dari jam kerja tersebut.

In [22]:
# Calculate pay based on working 40 hours
pay_fulltime = get_pay(40)
print(pay_fulltime)

528.0


Code di atas menggunakan fungsi `get_pay` untuk menghitung pembayaran setelah pajak bagi seseorang yang bekerja 40 jam. Hasil perhitungan tersebut disimpan dalam variabel `pay_fulltime`, dan kemudian nilai variabel tersebut dicetak menggunakan pernyataan `print`. Untuk memahami lebih lanjut, mari kita lihat proses perhitungannya:
1. Jumlah jam kerja adalah 40.
2. Pembayaran sebelum pajak dihitung dengan mengalikan jumlah jam (40) dengan tarif upah per jam ($15), sehingga `pay_pretax = 40 * 15 = 600`.
3. Pembayaran setelah pajak dihitung dengan mengalikan `pay_pretax` dengan nilai (1 - 0.12), yang mewakili 1 minus tarif pajak (12% atau 0.12), sehingga `pay_aftertax = 600 * (1 - 0.12) = 528`.

Jadi, hasil dari `print(pay_fulltime)` akan mencetak nilai 528, yang merupakan pembayaran setelah pajak bagi seseorang yang bekerja 40 jam dengan tarif upah per jam $15 dan tarif pajak sebesar 12%.

In [23]:
pay_parttime = get_pay(32)
print(pay_parttime)

422.4


Pembayaran setelah pajak untuk seseorang yang bekerja paruh waktu selama 32 jam menggunakan fungsi `get_pay`.

1. Jumlah jam kerja adalah 32.
2. Pembayaran sebelum pajak dihitung dengan mengalikan jumlah jam (32) dengan tarif upah per jam ($15), sehingga `pay_pretax = 32 * 15 = 480`.
3. Pembayaran setelah pajak dihitung dengan mengalikan `pay_pretax` dengan nilai (1 - 0.12), yang mewakili 1 minus tarif pajak (12% atau 0.12), sehingga `pay_aftertax = 480 * (1 - 0.12) = 422.40`.

Jadi, hasil dari `print(pay_parttime)` akan mencetak nilai 422.40, yang merupakan pembayaran setelah pajak bagi seseorang yang bekerja paruh waktu selama 32 jam dengan tarif upah per jam $15 dan tarif pajak sebesar 12%.

## Variable "scope"

In [24]:
print(pay_aftertax)

NameError: name 'pay_aftertax' is not defined


Error tersebut terjadi karena variabel `pay_aftertax` hanya didefinisikan di dalam lingkup fungsi `get_pay()`. Variabel tersebut tidak ada di luar fungsi tersebut atau di luar konteks di mana fungsi tersebut dipanggil.

## Functions with multiple arguments

In [ ]:
def get_pay_with_more_inputs(num_hours, hourly_wage, tax_bracket):
    # Pre-tax pay
    pay_pretax = num_hours * hourly_wage
    # After-tax pay
    pay_aftertax = pay_pretax * (1 - tax_bracket)
    return pay_aftertax

Fungsi di atas disebut `get_pay_with_more_inputs`, yang mengambil tiga argumen: `num_hours` (jumlah jam kerja), `hourly_wage` (tarif upah per jam), dan `tax_bracket` (tarif pajak dalam bentuk desimal). Fungsi ini menghitung pembayaran sebelum pajak dan setelah pajak berdasarkan input yang diberikan.

1. `pay_pretax` dihitung dengan mengalikan jumlah jam (`num_hours`) dengan tarif upah per jam (`hourly_wage`).
2. `pay_aftertax` dihitung dengan mengalikan `pay_pretax` dengan nilai (1 - `tax_bracket`), yang mewakili 1 dikurangi tarif pajak (dalam bentuk desimal).

Setelah itu, fungsi mengembalikan nilai `pay_aftertax`. Jadi, jika Anda memberikan jumlah jam kerja, tarif upah per jam, dan tarif pajak, fungsi akan mengembalikan jumlah pembayaran setelah pajak dari jam kerja tersebut.

In [ ]:
higher_pay_aftertax = get_pay_with_more_inputs(40, 24, .22)
print(higher_pay_aftertax)

748.8000000000001


Dalam pemanggilan fungsi `get_pay_with_more_inputs(40, 24, .22)`, kita memberikan tiga argumen:
1. `num_hours` = 40 (jumlah jam kerja).
2. `hourly_wage` = 24 (tarif upah per jam).
3. `tax_bracket` = 0.22 (tarif pajak dalam bentuk desimal, yang mewakili 22%).

Fungsi tersebut kemudian menghitung pembayaran setelah pajak berdasarkan argumen yang diberikan.
1. Pembayaran sebelum pajak dihitung dengan mengalikan jumlah jam (40) dengan tarif upah per jam (24), sehingga `pay_pretax = 40 * 24 = 960`.
2. Pembayaran setelah pajak dihitung dengan mengalikan `pay_pretax` dengan nilai (1 - 0.22), yang mewakili 1 dikurangi tarif pajak (22% dalam bentuk desimal), sehingga `pay_aftertax = 960 * (1 - 0.22) = 748.80`.

Jadi, `higher_pay_aftertax` akan memiliki nilai 748.80, yang merupakan pembayaran setelah pajak untuk seseorang yang bekerja 40 jam dengan tarif upah per jam $24 dan tarif pajak sebesar 22%.

In [ ]:
same_pay_fulltime = get_pay_with_more_inputs(40, 15, .12)
print(same_pay_fulltime)

528.0


Dalam pemanggilan fungsi `get_pay_with_more_inputs(40, 15, .12)`, kita memberikan tiga argumen:

1. `num_hours` = 40 (jumlah jam kerja).
2. `hourly_wage` = 15 (tarif upah per jam).
3. `tax_bracket` = 0.12 (tarif pajak dalam bentuk desimal, yang mewakili 12%).

Fungsi tersebut kemudian menghitung pembayaran setelah pajak berdasarkan argumen yang diberikan.

1. Pembayaran sebelum pajak dihitung dengan mengalikan jumlah jam (40) dengan tarif upah per jam (15), sehingga `pay_pretax = 40 * 15 = 600`.
2. Pembayaran setelah pajak dihitung dengan mengalikan `pay_pretax` dengan nilai (1 - 0.12), yang mewakili 1 dikurangi tarif pajak (12% dalam bentuk desimal), sehingga `pay_aftertax = 600 * (1 - 0.12) = 528`.

Jadi, `same_pay_fulltime` akan memiliki nilai 528, yang merupakan pembayaran setelah pajak untuk seseorang yang bekerja 40 jam dengan tarif upah per jam $15 dan tarif pajak sebesar 12%.

## Functions with no arguments

In [ ]:
# Define the function with no arguments and with no return
def print_hello():
    print("Hello, you!")
    print("Good morning!")
    
# Call the function
print_hello()

Hello, you!
Good morning!


Fungsi di atas, `print_hello`, tidak memiliki argumen dan tidak mengembalikan nilai apapun. Ini hanya mencetak dua pesan ke layar.

1. "Hello, you!"
2. "Good morning!"

Kemudian, fungsi tersebut dipanggil dengan menggunakan `print_hello()`, yang menyebabkan kedua pesan tersebut dicetak ke layar.